src/main/core/main.rs

```diff
@@ -0,0 +1,258 @@
+
+    // check if some log levels have been compiled out
+    let log_level = c_to_rust_log_level(log_level).unwrap();
+    if log_level > log::STATIC_MAX_LEVEL {
+        log::warn!(
+            "Log level set to {}, but messages higher than {} have been compiled out",
+            log_level,
+            log::STATIC_MAX_LEVEL,
+        );
+    }
+
+    // before we run the simluation, clean up any orphaned shared memory
+    unsafe { c::shmemcleanup_tryCleanup() };
+
+    // save the platform data required for CPU pinning
+    if config.experimental.use_cpu_pinning.unwrap() {
+        if unsafe { c::affinity_initPlatformInfo() } != 0 {
+            return Err(anyhow::anyhow!("Unable to initialize platform info"));
+        }
+    }
+
+    // raise fd soft limit to hard limit
+    raise_rlimit(libc::RLIMIT_NOFILE).context("Could not raise fd limit")?;
+
+    // raise number of processes/threads soft limit to hard limit
+    raise_rlimit(libc::RLIMIT_NPROC).context("Could not raise proc limit")?;
+
+    if config.experimental.use_sched_fifo.unwrap() {
+        set_sched_fifo().context("Could not set real-time scheduler mode to SCHED_FIFO")?;
+        log::debug!("Successfully set real-time scheduler mode to SCHED_FIFO");
+    }
+
+    // Disable address space layout randomization of processes forked from this
+    // one to improve determinism in cases when an executable under simulation
+    // branch on memory addresses.
+    disable_aslr().context("Could not disable plugin address space layout randomization")?;
+    log::debug!("ASLR disabled for processes forked from this parent process");
+
+    // check sidechannel mitigations
+    if unsafe { c::main_sidechannelMitigationsEnabled() } {
+        log::warn!(
+            "Speculative Store Bypass sidechannel mitigation is enabled (perhaps by seccomp?). \
+             This typically adds ~30%% performance overhead."
+        );
+    }
+
+    // log some information
+    unsafe { c::main_logBuildInfo() };
+    log_environment(args.clone());
+
+    log::debug!("Startup checks passed, we are ready to start the simulation");
+
+    // allow gdb to attach before starting the simulation
+    if options.gdb {
+        pause_for_gdb_attach().context("Could not pause shadow to allow gdb to attach")?;
+    }
+
+    // allocate and initialize our main simulation driver
+    let controller = unsafe { c::controller_new(&config as *const _) };
+    assert!(!controller.is_null());
+
+    // run the simulation
+    let rv = unsafe { c::controller_run(controller) };
+    unsafe { c::controller_free(controller) };
+    std::mem::drop(controller);
+
+    if rv != 0 {
+        return Err(anyhow::anyhow!("Controller exited with code {}", rv));
+    }
+
+    Ok(())
+}
+
+fn pause_for_gdb_attach() -> anyhow::Result<()> {
+    let pid = nix::unistd::getpid();
+    log::info!(
+        "Pausing with SIGTSTP to enable debugger attachment (pid {})",
```

Comment:
SIGSTOP

Reply:
Oops, good catch! I changed the `raise(SIGSTOP)` to `raise(SIGTSTP)` instead to match the original C code.